File: NoroffAlumni/src/main/java/no/experisAcadmey/trondheim/NoroffAlumni/services/TopicService.java
package no.experisAcadmey.trondheim.NoroffAlumni.services;
import no.experisAcadmey.trondheim.NoroffAlumni.exceptions.TopicExistException;
import no.experisAcadmey.trondheim.NoroffAlumni.models.DTOs.topicDTOs.NewTopic;
import no.experisAcadmey.trondheim.NoroffAlumni.models.Post;
import no.experisAcadmey.trondheim.NoroffAlumni.repositories.PostRepository;
import org.springframework.data.domain.Pageable;

import no.experisAcadmey.trondheim.NoroffAlumni.exceptions.TopicNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.*;

import no.experisAcadmey.trondheim.NoroffAlumni.repositories.TopicRepository;
import no.experisAcadmey.trondheim.NoroffAlumni.models.Topic;


@Service
public class TopicService {
   

   @Autowired
   private TopicRepository topicRepository;
   @Autowired
   private UserService userService;
   @Autowired
   private PostRepository postRepository;


   /**
    * Retrieves a list of Topics with the possibility
    * of search, and pagination
    * @param searchWord word to search for
    * @param page page number
    * @param pageSize page size
    * @return list of topics
    */
   public List<Topic> getTopics(Optional<String> searchWord, Optional<Integer> page, Optional<Integer> pageSize){
      List<Topic> foundTopics = new ArrayList<>();
      if(searchWord.isPresent() && page.isPresent() && pageSize.isPresent()){
         Pageable pageRequest = PageRequest.of(page.get(), pageSize.get(), Sort.by("name").ascending());
         foundTopics = topicRepository.findByNameContainingIgnoreCase(searchWord.get(),pageRequest).toList();
      }else if(searchWord.isPresent()){
         foundTopics = topicRepository.findByNameContainingIgnoreCase(searchWord.get());
      }else {
         Pageable pageRequest = PageRequest.of(page.orElse(0), pageSize.orElse(10), Sort.by("name").ascending());
         foundTopics = topicRepository.findAll(pageRequest).toList();
      }

      if(foundTopics.size() == 0) throw new TopicNotFoundException(searchWord);

      //throw new NotFoundException();
      return foundTopics;
   }

   /**
    * Retrieves all topics that a user is subscribed to
    * @return List of subscribed topics for user
    */
   public List<Topic> getSubscribedTopics() {
      return userService.getCurrentUser().getTopics().stream().toList();
   }

   /**
    * Retrieves a specific topic based on Id
    *
    * @param topicId id of the topic to find
    * @return found topic
    */
   public Topic getTopic(Long topicId){
      return topicRepository.findById(topicId).orElseThrow(TopicNotFoundException::new);
   }

   /**
    * Retrieves a specific topic based on name
    *
    * @param topicName Name of the topic to find
    * @return found topic
    */
   public Topic getTopicByName(String topicName){
      return topicRepository.findByName(topicName).orElseThrow(TopicNotFoundException::new);
   }

   /**
    * Creates a new Topic.
    *
    * @param newTopic data of the new topic.
    * @return id of the new topic.
    * @throws TopicExistException thrown if the topic already exists.
    */
   public Long createTopic(NewTopic newTopic) throws TopicExistException {
      if(topicRepository.findByName(newTopic.getName()).isPresent()) throw new TopicExistException();

      Topic topic = new Topic();
      topic.setName(newTopic.getName());
      topic.setDescription(newTopic.getDescription());
      topic.setSubscribers(Set.of(userService.getCurrentUser()));
      topic.setPosts(new ArrayList<>());

      topicRepository.save(topic);

      return topic.getId();
   }

   /**
    * Join topic based on current user and topic id.
    * @param topicId id of the topic to join.
    */
   public Topic joinTopic(Long topicId){
      Topic topic = topicRepository.findById(topicId).orElseThrow(TopicNotFoundException::new);
      topic.addSubscriber(userService.getCurrentUser());
      return topicRepository.save(topic);
   }

   public void leaveTopic(Long topicId){
      Topic topic = topicRepository.findById(topicId).orElseThrow(TopicNotFoundException::new);
      topic.removeSubscriber(userService.getCurrentUser());
      topicRepository.save(topic);
   }

   public List<Post> findTopicPosts(Long topicId,Integer pageNumber,Integer pageSize){
      Pageable pageRequest = PageRequest.of(pageNumber != null ? pageNumber : 0,pageSize != null? pageSize : 10 );
      List<Post> foundPosts = postRepository.findAllByTargetTopicsIdOrderByLastUpdated(topicId,pageRequest).toList();
      return removeDirectMessages(foundPosts);
   }

   public List<Post> removeDirectMessages(List<Post> posts){
      List<Post> result = new ArrayList<>();
      for(Post post : posts) {
         if(post.getTargetUser() == null){
            result.add(post);
         }
      }
      return result;
   }
}

File: noroff-alumni-frontend/src/services/topicService.ts
import NewTopic from "../models/NewTopic";
import api from "./api";

class TopicService {
  async getTopics(pageNum: number, pageSize: number) {
    return await api
      .get("/topic", {
        params: {
          page: pageNum,
          pageSize: pageSize,
        },
      })
      .then((response) => {
        return response.data;
      });
  }

  async getSubscribedTopics(){
    return await api.get("/topic/subscribed")
        .then(response => response.data);
  }

  async joinTopic(topicId: number) {
    return await api.post("/topic/" + topicId + "/join").then((response) => {
      return response.data;
    });
  }

  async getTopic(topicId: number) {
    return await api.get("/topic/" + topicId).then((response) => {
      return response.data;
    });
  }

  async getTopicPosts(topicId: number, pageNum: number, pageSize: number) {
    return await api
      .get("/topic/" + topicId+"/posts", {
        params: {
          page: pageNum,
          pageSize: pageSize,
        },
      })
      .then((response) => {
        return response.data;
      });
  }

  async searchTopics(searchWord: string, pageNum: number, pageSize: number) {
    return await api
      .get("/topic", {
        params: {
          page: pageNum,
          pageSize: pageSize,
          search: searchWord,
        },
      })
      .then((response) => {
        return response.data;
      });
  }

  async createTopic(newTopic: NewTopic) {
    await api.post("/topic", newTopic);
  }

  async leaveTopic(topicId:number){
    return await api.post("/topic/" + topicId + "/leave").then((response) => {
      return response.data;
    });
  }
}
export default new TopicService();
